File: benchmarks/bench.lua
require "socket"
require "haml"

local n = 5000

local template = [=[
!!! html
%html
  %head
    %title Test
  %body
    %h1 simple markup
    %div#content
    %ul
      - for _, letter in ipairs({"a", "b", "c", "d", "e", "f", "g"}) do
        %li= letter
]=]

local start = socket.gettime()
for i = 1,n do
  local engine = haml.new()
  local html = engine:render(template)
end
local done = socket.gettime()

print "Compile and render:"
print(("%s seconds"):format(done - start))

local engine        = haml.new()
local phrases       = engine:parse(template)
local compiled      = engine:compile(phrases)
local haml_renderer = require "haml.renderer"
local renderer      = haml_renderer.new(compiled)

local start = socket.gettime()
for i = 1,n do
  renderer:render(compiled)
end
local done = socket.gettime()

print "Render:"
print(("%s seconds"):format(done - start))

File: benchmarks/bench.rb
require "benchmark"
require "haml"

n = 5000

haml = '!!! html
%html
  %head
    %title Test
  %body
    %h1 simple markup
    %div#content
    %ul
    - ["a", "b", "c", "d", "e", "f", "g"].each do |letter|
      %li= letter'

compiled = Haml::Engine.new(haml, :format => :html5, :ugly => true)


Benchmark.bmbm do |bench|
  bench.report("haml #{Haml::VERSION} - compile & render") do
    for i in 0..n do
      Haml::Engine.new(haml, :format => :html5, :ugly => true).render
    end
  end
  bench.report("haml #{Haml::VERSION} - render") do
    for i in 0..n do
      compiled.render
    end
  end
end